tests: posix: common: Rework comparison of REALTIME and MONOTONIC clock

ABOSTM opened the pull request:
tests: posix: common: Rework comparison of REALTIME and MONOLITIC clock.
Between both clock_gettime(), time may have been increased by 1ms (precision of k_uptime_get()).
So both time may not be equal.
This could imply a rollover of nsec counter and increment number of seconds.
Rework assert check to take those cases into account.

Fixes #25453

pabigot reviewed (dismissed):
I'm unable to make this test fail on this platform, though I agree it's technically possible if something failed in the tick alignment.  Perhaps there's something else going on.

I think your solution is correct, though something like this might be more clear:
```
diff --git a/tests/posix/common/src/clock.c b/tests/posix/common/src/clock.c
index a477ceb81d..eec10fa983 100644
--- a/tests/posix/common/src/clock.c
+++ b/tests/posix/common/src/clock.c
@@ -54,6 +54,8 @@ void test_posix_realtime(void)
        int ret;
        struct timespec rts, mts;
        struct timeval tv;
+       u64_t mtns;
+       u64_t rtns;
 
        /* Minimal sleep to align us to the next tick interval. This
         * helps with a case that 2 consecutive calls to clock_gettime()
@@ -70,8 +72,14 @@ void test_posix_realtime(void)
        ret = clock_gettime(CLOCK_REALTIME, &rts);
        zassert_equal(ret, 0, "Fail to get realtime clock");
 
-       zassert_equal(rts.tv_sec, mts.tv_sec, "Seconds not equal");
-       zassert_equal(rts.tv_nsec, mts.tv_nsec, "Nanoseconds not equal");
+       /* Error if rts is more than 1 ms different from mts.  (They
+        * should be equal, but if the sync k_usleep() above didn't
+        * succeed a single ms rollover of the underlying k_uptime
+        * clock might occur between the two calls). */
+       mtns = mts.tv_sec * NSEC_PER_SEC + mts.tv_nsec;
+       rtns = rts.tv_sec * NSEC_PER_SEC + rts.tv_nsec;
+       zassert_true((rtns - mtns) <= (NSEC_PER_SEC / MSEC_PER_SEC),
+                    "MONOTONIC !~ REALTIME");
 
        /* Set a particular time.  In this case, the output of:
         * `date +%s -d 2018-01-01T15:45:01Z`
```

pfalcon commented:
> test: posix: common: Rework comparison of REALTIM

Please fix typo in the commit title.

ABOSTM commented:
> Please fix typo in the commit title.

Done

pfalcon commented on tests/posix/common/src/clock.c:
```diff
@@ -70,8 +71,31 @@ void test_posix_realtime(void)
 
-	zassert_equal(rts.tv_sec, mts.tv_sec, "Seconds not equal");
-	zassert_equal(rts.tv_nsec, mts.tv_nsec, "Nanoseconds not equal");
+	/* Between both clock_gettime(), time may have been increased by 1ms
```
It's nice reminder that the time keeping has granularity, besides precision. I.e., the fact that somewhere time is kept in nanoseconds in milliseconds, doesn't mean it increments in such units. But in that regard, I don't think we can guarantee a particular granularity across Zephyr architectures and versions. For example, I remember seeing granularity of 10ms. So, I'd suggest to remove references to "1ms" and use more generic wording, and then apparently also introduce a symbolic constant for granularity (and be ready to bump it up, though it can start with 1ms).

pfalcon commented on tests/posix/common/src/clock.c:
```diff
@@ -70,8 +71,31 @@ void test_posix_realtime(void)
+		}
+		zassert_equal(rts.tv_nsec, mts_tvnsec_plus1,
+			"Nanoseconds neither equal, nor differ by 1");
+	} else {
```
I'd suggest to remove this empty else. It doesn't add clarity, just make an already big `if` even more unwieldy.

pfalcon reviewed (commented):
Just a week ago I had a similar situation with this "test case". I had a choice to either remove it (see comment at the beginning of this function), or keep by finding a simple workaround, and with @andyross help, such a simple workaround was found: https://github.com/zephyrproject-rtos/zephyr/commit/b07489614f44ffa11973e60a8b9ff957876cb1d8 .

If there was a choice between adding such a pretty big code chunk and removing flaky comparison, I'm not sure what my decision would be, or rather, I think I'd prefer to just remove it. Maybe, this impression is based on pretty complicated (relatively, of course) "if" condition added. Maybe, if it worked by converting time from vector (seconds, nanos) into scalar form, and diffing scalar values, it would give different impression (maybe not).

All in all, I'd leave the decision to @andyross, if he finds creeping workarounds of this matter acceptable, then it's ok.

andyross commented on tests/posix/common/src/clock.c:
```diff
@@ -70,8 +71,31 @@ void test_posix_realtime(void)
+	 */
+	if ((rts.tv_nsec != mts.tv_nsec) || (rts.tv_sec != mts.tv_sec)) {
+		/* If nsec or sec are not equal, then rts should be equal to
+		 * mts + 1ms
```
That comment doesn't sound like correct logic to me.  Nothing in either Zephyr or POSIX guarantees that two consecutive times differ by exactly 1ms.

It also doesn't seem like that's the logic implemented below, which is looking for a delta of exactly one **nanosecond**!  And I can 100% guarantee that Zephyr won't produce that -- it's fully five orders of magnitude more precise than the highest precision tick rate on any existing system.

pabigot replied:
This test is verifying the behavior of the Zephyr POSIX implementation of `clock_gettime()`, which returns the value of `k_uptime_get()` for monotonic and `k_uptime_get()` plus a base offset for "realtime".  So unless/until that changes it will always increment in multiples of 1 ms.

None of which is against the point you or @pfalcon raise about this test's inherent weaknesses, just that it is meeting the minimum expectations for what it's intended to verify.

pfalcon replied:
> So unless/until that changes it will always increment in multiples of 1 ms.

The point, that it may, and **does**, increment in **multiples** of 1ms. Many people (re)discover again and again that e.g. on QEMU, time granularity is 10ms (see e.g. cb2bfcb9c0e73e271f45a3d6ac8cc9718810e3c9 and gazillion of other commits (I looked for mine and saw 3-4 from other people on the matter)). Thus, both saying and actually testing that difference can be 1ms is of little help.

pfalcon replied:
Here's for example how I do it: https://github.com/zephyrproject-rtos/zephyr/blob/21358baa72cea9c23be57444eb91444774842f97/tests/net/socket/select/src/main.c#L26 (again, defining a symbolic fuzz factor, and set it conservatively ("crudest" known (at least, to me) time granularity is 10ms)).

pabigot replied:
As the test describes in the comment it also uses `k_usleep(1)` to align to the start of a tick, so that the two times will be equal even on QEMU within the limits of the test's goals and context.  In fact, there's only one platform where this patch is asserted to be relevant, and I can't replicate the failure on that platform.

Really what this test is trying to do is to confirm, through a secondary effect, that the offset of `CLOCK_REALTIME` from `CLOCK_MONOTONIC` is zero.  We could just remove the test instead, as even a 1 ms slop as added here invalidates the purpose of the test.

In any case, I confirm that this is a working solution to #25453 if in fact that issue is reproducible.  The logic does not look for a delta of 1 ns, but I agree it's too complex and the name `mts_tvnsec_plus1` is misleading.  So I've retracted my approval, though I'd still approve the alternative implementation I suggested.

ABOSTM replied:
Sure, I reproduced the issue systematicvally (on a certain commit). And yes my solution solves the issue.
But I understand it is a bit complex, and I will give a try with your suggested implementation.

pfalcon replied:
> As the test describes in the comment it also uses k_usleep(1) to align to the start of a tick,

That k_usleep() and comment was added by me 10 days ago on @andyross suggestion, on seeing a failing case very similar to #25453 (that case was also on real hw, not qemu: https://github.com/zephyrproject-rtos/zephyr/issues/25017).

The more strange that this "alignment to a start of a ticket" didn't work and 10 days later, a similar issue pops up on another hw.


I don't comment on other points, because they're similar to what I already said in my review comment: https://github.com/zephyrproject-rtos/zephyr/pull/25454#pullrequestreview-415200109

The only new thought I can bring, is that if there's a desire to keep applying small-scale (1-2 lines, not counting comments) workarounds, then one can wrap 2 clock_gettime() calls in irq_lock(). We still will need to remove that comparison eventually (e.g. when CLOCK_REALTIME will be actually RTC-backed). I didn't remove it 10 days ago only because I consider that we lack a lot of test coverage, so removing tests added by somebody else isn't a good way to encourage people to write more. (Otherwise, we can "resolve" an issue with any test by just removing it.)

pfalcon replied:
> I understand it is a bit complex,

It's not complex, it's confusing. It's unclear why workarounds applied so far don't work consistently across hw, and we need to apply more (and then it would make sense to apply such which wouldn't break again in a week).

andyross reviewed (changes requested):
Yeah, this doesn't look like it does what you want.   The root problem here is that you can't lock against the passage of time in the real world, so two calls to k_uptime_get() (or the POSIX wrappers here, in this particular case) aren't guaranteed to give you the same value no matter how hard you try.

If you want to allow slop in the comparison, then just convert both values to a signed 64 bit count of nanoseconds, subtract, and test the absolute value vs. a threshold.  Simpler and more correct.

ABOSTM commented:
I could implement pabigot proposal, but even this solution doesn't seem to reach a consensus.
@carlescufi how can we proceed for this PR, (which is finally not STM32 specific) ?

carlescufi commented:
> I could implement pabigot proposal, but even this solution doesn't seem to reach a consensus.
> @carlescufi how can we proceed for this PR, (which is finally not STM32 specific) ?

I suggest we tag this as dev-review and discuss it today.

ABOSTM commented:
yes

galak commented:
Dev-review (May 28): will try implementing @pabigot proposed code change and push that if it works.

ABOSTM committed:
tests: posix: common: Rework comparison of REALTIME and MONOTONIC clock

Between both clock_gettime(), time may have been increased by 1ms
(precision of k_uptime_get()). So allow 1s toleance in  both time
comparison.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM commented:
PR updated with @pabigot code.
It fixes original issue #25453

pabigot commented:
This is really interesting.

I can reproduce this on wb55 at commit: a0ae53410eef31d1773dbfc5a3a6591e45dee11c

It works fine at the next commit: ef25b864e28c3b36b6a98ce468f022d2412e20e4

The difference?

The version string for the first commit is:

    v2.3.0-rc1-99-ga0ae53410eef

The version string for the second commit is:

    v2.3.0-rc1-100-gef25b864e28c

This ends up shifting the address for various read-only strings by one byte.  There are no differences to the generated instructions, just the address values they load.  I see no differences outside the `.rodata.*.str1.*` section.

There's something funky going on.  My guess is there's a misalignment in the failing case that causes a stall during which allows an extra tick to occur.

FYI @andyross @pfalcon.  The relaxed check here seems justified if the test is this fragile.

pfalcon commented:
@pabigot: That's interesting indeed, if it's 100% reproducible across multiple runs. Note that I'm ok with any fix here, as long as @andyross is ok.

ABOSTM commented:
Rebase, with uint64_t instead of u64_t

ABOSTM commented:
Is there any reason to not merge this PR ?

pfalcon commented:
@ABOSTM: This PR cannot be merged, because it has -1 vote. If you addressed comments of such a review, you would need to ping the review author to re-review.

ABOSTM commented:
@andyross can you please review ?

ABOSTM commented:
@andyross can you please review ?

ABOSTM commented:
@andyross can you please review ?

ABOSTM commented:
Closing; #28516 has been selected instead.
